[SYCL] getDeviceLambdaManglingNumber() const: Assertion `isLambda() && "Not a lambda closure type!"' failed.

Building the example below using the SYCL compiler with Cuda support shows the error:

/path/to/clang/lib/AST/DeclCXX.cpp:1613: unsigned int clang::CXXRecordDecl::getDeviceLambdaManglingNumber() const: Assertion `isLambda() && "Not a lambda closure type!"'

https://github.com/zjin-lcf/HeCBench/tree/master/vmc-sycl

I have created a fix in #7927, although, it still won't help to compile example. This is because the example passes unnamed struct to `handler::copy` here https://github.com/zjin-lcf/HeCBench/blob/master/vmc-sycl/vmc2.cpp#L310 . Under the hood `handler::copy` creates a kernel and it is implemented in a way so the type of destination pointer ends up being a part of kernel name. Unnamed struct causes this name to be invalid and triggers the following error:
```
llvm/build/bin/../include/sycl/handler.hpp:1279:59: error: unnamed type '__copyAcc2Ptr<float, (unnamed struct at vmc2.cpp:306:5), 1, sycl::ac
cess::mode::read, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>' is invalid; provide a kernel name, or use '-fsycl-unnamed-lambda' to enable unna
med kernel lambdas
      h->kernel_parallel_for<TypesToForward..., Props...>(Args...);
                                                          ^
llvm/build/bin/../include/sycl/handler.hpp:1356:29: note: in instantiation of function template specialization 'sycl::handler::KernelProperti
esUnpackerImpl<>::kernel_parallel_for_unpack<__copyAcc2Ptr<float, (unnamed struct at vmc2.cpp:306:5), 1, sycl::access::mode::read, sycl::access::target::global_buffer, sycl
::access::placeholder::false_t>, sycl::item<1, true>, (lambda at llvm/build/bin/../include/sycl/handler.hpp:797:16), (lambda at /iusers/mpodc
his/source/syclos/llvm/build/bin/../include/sycl/handler.hpp:797:16)>' requested here
          Unpacker.template kernel_parallel_for_unpack<KernelName, ElementType,
                            ^
llvm/build/bin/../include/sycl/handler.hpp:1352:5: note: in instantiation of function template specialization 'sycl::handler::unpack<(lambda
at llvm/build/bin/../include/sycl/handler.hpp:797:16), sycl::ext::oneapi::experimental::properties<std::tuple<>>, false, (lambda at /iusers/m
podchis/source/syclos/llvm/build/bin/../include/sycl/handler.hpp:1355:21)>' requested here
    unpack<KernelType, PropertiesT,
    ^
llvm/build/bin/../include/sycl/handler.hpp:1004:7: note: in instantiation of function template specialization 'sycl::handler::kernel_parallel
_for_wrapper<__copyAcc2Ptr<float, (unnamed struct at vmc2.cpp:306:5), 1, sycl::access::mode::read, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>,
 sycl::item<1, true>, (lambda at llvm/build/bin/../include/sycl/handler.hpp:797:16), sycl::ext::oneapi::experimental::properties<std::tuple<>
>>' requested here
      kernel_parallel_for_wrapper<NameT, TransformedArgType, KernelType,
      ^
llvm/build/bin/../include/sycl/handler.hpp:1996:5: note: in instantiation of function template specialization 'sycl::handler::parallel_for_la
mbda_impl<__copyAcc2Ptr<float, (unnamed struct at vmc2.cpp:306:5), 1, sycl::access::mode::read, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>, (l
ambda at llvm/build/bin/../include/sycl/handler.hpp:797:16), 1, sycl::ext::oneapi::experimental::properties<std::tuple<>>>' requested here
    parallel_for_lambda_impl<KernelName, KernelType, 1, PropertiesT>(
    ^
llvm/build/bin/../include/sycl/handler.hpp:2046:5: note: in instantiation of function template specialization 'sycl::handler::parallel_for<__
copyAcc2Ptr<float, (unnamed struct at vmc2.cpp:306:5), 1, sycl::access::mode::read, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>, (lambda at /iu
sers/mpodchis/source/syclos/llvm/build/bin/../include/sycl/handler.hpp:797:16), sycl::ext::oneapi::experimental::properties<std::tuple<>>>' requested here
    parallel_for<KernelName>(
    ^
llvm/build/bin/../include/sycl/handler.hpp:795:5: note: in instantiation of function template specialization 'sycl::handler::parallel_for<__c
opyAcc2Ptr<float, (unnamed struct at vmc2.cpp:306:5), 1, sycl::access::mode::read, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>, 1, (lambda at /
iusers/mpodchis/source/syclos/llvm/build/bin/../include/sycl/handler.hpp:797:16)>' requested here
    parallel_for<
    ^
```
The obvious w/a is to not use unnamed struct.
The proper solution lays inside SYCL headers implementation, I believe. I suggest filing a separate issue for that.

--- a/clang/lib/AST/Expr.cpp
+++ b/clang/lib/AST/Expr.cpp
@@ -644,7 +644,8 @@ std::string SYCLUniqueStableNameExpr::ComputeName(ASTContext &Context) const {
 static llvm::Optional<unsigned>
 UniqueStableNameDiscriminator(ASTContext &, const NamedDecl *ND) {
   if (const auto *RD = dyn_cast<CXXRecordDecl>(ND))
-    return RD->getDeviceLambdaManglingNumber();
+    if (RD->isLambda())
+      return RD->getDeviceLambdaManglingNumber();
   return llvm::None;
 }
 

--- a/clang/test/SemaSYCL/unique-stable-name-anon-struct-crash.cpp
+++ b/clang/test/SemaSYCL/unique-stable-name-anon-struct-crash.cpp
@@ -1,0 +1,16 @@
+// RUN: %clang_cc1 %s %s -std=c++17 -triple x86_64-linux-gnu -Wno-sycl-2020-compat -fsycl-is-device -verify -fsyntax-only -Wno-unused
+
+// This would crash due to UniqueStableNameDiscriminator expecting only lambdas
+// as unnamed records.
+// expected-no-diagnostics
+
+template <typename T1, typename T2>
+class TC {};
+
+template <typename T1, typename T2> void TFoo(T1 _T1, T2 _T2) {
+ __builtin_sycl_unique_stable_name(TC<T1, T2>);
+}
+void use() {
+  struct {float r1, r2, r12; } s;
+  TFoo(1, s);
+}

